sql/src/main/java/io/crate/operation/collect/OrderedLuceneDocCollector.java

```diff
@@ -0,0 +1,273 @@
+                return;
+            }
+            internalCollectContext.lastCollected = scoreDoc;
+            internalCollectContext.topFieldDocs = null;
+        }
+    }
+
+    @Override
+    protected void buildContext(Query query) {
+        if (internalCollectContext == null) {
+            Sort sort = LuceneSortGenerator.generateLuceneSort(searchContext, orderBy, inputSymbolVisitor);
+            internalCollectContext = new InternalCollectContext(query, orderBy, sort);
+        }
+    }
+
+    @Override
+    protected boolean skipDoc(int doc) {
+        return false;
+    }
+
+    @Override
+    protected void postCollectDoc(int doc) {
+        // do nothing, all done in searchAndCollect()
+    }
+
+    @Override
+    protected void skipSegmentReader() {
+        // do nothing, all done in searchAndCollect()
+    }
+
+    public void setNextOrderByValues(ScoreDoc scoreDoc) {
+        for (OrderByCollectorExpression expr : orderByCollectorExpressions) {
+            expr.setNextFieldDoc((FieldDoc) scoreDoc);
+        }
+    }
+
+    private int batchSize() {
+        return limit == null ? batchSizeHint : Math.min(batchSizeHint, limit - rowCount);
+    }
+
+    protected static class InternalCollectContext extends LuceneDocCollector.InternalCollectContext {
+
+        private final DummyScorer scorer = new DummyScorer();
+
+        @Nullable
+        private final Sort sort;
+        @Nullable private final OrderBy orderBy;
+
+        private TopFieldDocs topFieldDocs;
+        private int topFieldPosition = 0;
+
+        public InternalCollectContext(Query query, @Nullable OrderBy orderBy, @Nullable Sort sort) {
+            super(query);
+            this.sort = sort;
+            this.orderBy = orderBy;
+        }
+
+        public Query buildQuery() {
+            if (lastCollected != null) {
+                Query nextPageQuery = nextPageQuery((FieldDoc) lastCollected);
+                if (nextPageQuery != null) {
+                    BooleanQuery searchAfterQuery = new BooleanQuery();
+                    searchAfterQuery.add(query, BooleanClause.Occur.MUST);
+                    searchAfterQuery.add(nextPageQuery, BooleanClause.Occur.MUST_NOT);
+                    return searchAfterQuery;
+                }
+            }
+            return query;
+        }
+
+        @Nullable
+        private Query nextPageQuery(FieldDoc lastCollected) {
+            if (orderBy == null) {
+                return null;
+            }
+            BooleanQuery query = new BooleanQuery();
+            for (int i = 0; i < orderBy.orderBySymbols().size(); i++) {
+                Symbol order = orderBy.orderBySymbols().get(i);
+                Object value = lastCollected.fields[i];
+                // only filter for null values if nulls last
+                if (order instanceof Reference && (value != null || !orderBy.nullsFirst()[i])) {
+                    QueryBuilderHelper helper = QueryBuilderHelper.forType(order.valueType());
```

Comment:
would probably make sense to do something of this builder stuff outside the page loop